web/static/js/content_types/image_tray.js

```diff
@@ -56,35 +65,68 @@ class ImageTray extends React.Component {
     })
   }
 
-  onUpload (err, metadata) {
+  onOspryUpload (err, metadata) {
     if (err) return window.alert(err)
     this.setState({ url: metadata.httpsURL })
   }
 
   uploadFile (e) {
     this.setState({ fileName: e.target.files[0].name })
 
+    e.preventDefault()
+    Net.post('/thesis/files/upload', e.target.form, 'form')
+    .then((response) => {
+      if (response.path.length > 0) this.setState({ url: response.path })
+    })
+    .catch((error) => { window.alert(error) })
+  }
+
+  uploadFileToOspry (e) {
+    this.setState({ fileName: e.target.files[0].name })
+
     const ospry = new Ospry(this.props.ospryPublicKey)
 
     e.preventDefault()
     const form = e.target
     ospry.up({
       form: form,
-      imageReady: this.onUpload
+      imageReady: this.onOspryUpload
     })
   }
 
   previewImageStyle () {
     return {backgroundImage: `url(${this.state.url})`}
   }
 
+  renderForm () {
+    if (this.props.ospryPublicKey) {
+      return this.renderOspryForm()
+    } else if (this.props.fileUploader) {
+      return this.renderUploaderForm()
+    }
+  }
```

Comment:
Wonder if we could handle Ospry server-side, entirely, and send through the normal upload form.